src/distributed_join.cuh

```diff
@@ -152,6 +163,219 @@ void all_to_all_comm(cudf::table_view input,
+    CUDA_RT_CALL(cudaStreamSynchronize(cudaStreamDefault));
+  }
+};
+
+/**
+ * All-to-all communication of a single batch, with data compression.
+ *
+ * After the all-to-all communication is completed, the data can be decompressed with
+ * *all_to_all_comm_decompression*.
+ *
+ * Note: If the communicator supports grouping by batches, this call is nonblocking and should
+ * be enclosed by `communicator->start()` and `communicator->stop()`.
+ *
+ * @param[in] input Table to be communicated.
+ * @param[in] send_offset Vector of length mpi_size + 1 such that `send_offset[i+1] -
+ * send_offset[i]` is the number of elements sent from the current rank to rank i during the
+ * all-to-all communication.
+ * @param[out] compressed_input Vector of length number of columns, where each element holds the
+ * compressed data of a column to be sent to remote GPUs. This argument does not need to be
+ * preallocated, but the user of this function needs to keep it alive until the communication is
+ * finished.
+ * @param[out] compressed_output Vector of length number of columns, where each element holds the
+ * compressed data of a column received from each remote GPUs. This argument does not need to be
+ * preallocated.
+ * @param[out] compressed_recv_offset `compressed_recv_offset[i,j]` represents the start index of
+ * compressed data in `compressed_output[i]` received from rank `j`.
+ * @param[in] communicator An instance of `Communicator` used for communication.
+ * @param[in] include_self If true, this function will send the partition destined to the current
+ * rank.
+ */
+void all_to_all_comm_with_compression(cudf::table_view input,
+                                      vector<cudf::size_type> const &send_offset,
+                                      vector<rmm::device_buffer> &compressed_input,
+                                      vector<rmm::device_buffer> &compressed_output,
+                                      vector<vector<int64_t>> &compressed_recv_offset,
+                                      Communicator *communicator,
+                                      bool include_self = true)
+{
+  cudf::size_type ncols = input.num_columns();
+  compressed_input.resize(ncols);
+  compressed_output.resize(ncols);
+  compressed_recv_offset.resize(ncols);
+
+  int mpi_rank = communicator->mpi_rank;
+  int mpi_size = communicator->mpi_size;
+
+  for (cudf::size_type icol = 0; icol < ncols; icol++) {
+    cudf::column_view column   = input.column(icol);
+    cudf::data_type dtype      = column.type();
+    cudf::size_type dtype_size = cudf::size_of(dtype);
+
+    vector<rmm::device_buffer> compressed_column(mpi_size);
+    vector<size_t> compressed_column_size(mpi_size, 0);
+
+    // compress each partition in the column separately and store the result in *compressed_column*
+    for (int irank = 0; irank < mpi_size; irank++) {
+      if (!include_self && irank == mpi_rank) continue;
+
+      cudf::type_dispatcher(
```

Comment:
If we have a lot of partitions or small partitions, we might want to rewrite this section to 1) utilize batched LZ4 compressor; 2) enable better pipelining of cascaded compressor, so that we dispatch all async kernels independently without doing stream sync after each one. Just something to think about as future optimizations (no need to do it in this PR).

Reply:
Issue is raised internally.